verilog/analysis/symbol_table_test.cc

```diff
@@ -2183,6 +2183,138 @@ TEST(BuildSymbolTableTest, ModuleDeclarationLocalsDependOnParameter) {
   }
 }
 
+TEST(BuildSymbolTableTest, ModuleSingleImplicitDeclaration) {
+  TestVerilogSourceFile src("foo.sv",
+                            "module m;"
+                            "assign a = 1'b0;"
+                            "endmodule\n");
+  const auto status = src.Parse();
+  ASSERT_TRUE(status.ok()) << status.message();
+  SymbolTable symbol_table(nullptr);
+  const SymbolTableNode& root_symbol(symbol_table.Root());
+
+  const auto build_diagnostics = BuildSymbolTable(src, &symbol_table);
+  EXPECT_EMPTY_STATUSES(build_diagnostics);
+
+  MUST_ASSIGN_LOOKUP_SYMBOL(module_m, root_symbol, "m");
+  EXPECT_EQ(module_m_info.metatype, SymbolMetaType::kModule);
+  EXPECT_EQ(module_m_info.file_origin, &src);
+  EXPECT_EQ(module_m_info.declared_type.syntax_origin,
+            nullptr);  // there is no module meta-type
+
+  MUST_ASSIGN_LOOKUP_SYMBOL(a_variable, module_m, "a");
+  EXPECT_EQ(a_variable_info.metatype, SymbolMetaType::kDataNetVariableInstance);
+  const ReferenceComponentNode* a_type_ref =
+      a_variable_info.declared_type.user_defined_type;
+  EXPECT_EQ(a_type_ref, nullptr);  // implicit type is primitive type
+
+  EXPECT_EQ(module_m_info.local_references_to_bind.size(), 1);
+  const auto ref_map(module_m_info.LocalReferencesMapViewForTesting());
+
+  ASSIGN_MUST_FIND(a_refs, ref_map, "a");
+  ASSERT_EQ(a_refs.size(), 1);  // all references to "N" parameter
+  for (const auto& a_ref : a_refs) {
+    const ReferenceComponent& a_ref_comp(a_ref->components->Value());
+    EXPECT_EQ(a_ref_comp.ref_type, ReferenceType::kUnqualified);
+    EXPECT_EQ(a_ref_comp.required_metatype, SymbolMetaType::kUnspecified);
+    EXPECT_EQ(a_ref_comp.identifier, "a");
+    EXPECT_EQ(a_ref_comp.resolved_symbol, &a_variable);  // pre-resolved
+  }
+
+  {
+    std::vector<absl::Status> resolve_diagnostics;
+    // Resolve mustn't break anything
+    symbol_table.Resolve(&resolve_diagnostics);
+    EXPECT_EMPTY_STATUSES(resolve_diagnostics);
+
+    // All references to "a" resolved.
+    for (const auto& a_ref : a_refs) {
+      const ReferenceComponent& a_ref_comp(a_ref->components->Value());
+      EXPECT_EQ(a_ref_comp.resolved_symbol,
+                &a_variable);  // resolved successfully
+    }
+  }
+}
+
+TEST(BuildSymbolTableTest, ModuleReferenceToImplicitDeclaration) {
+  TestVerilogSourceFile src("foo.sv",
+                            "module m;"
+                            "assign a = 1'b0;"
+                            "assign a = 1'b1;"
+                            "endmodule\n");
+  const auto status = src.Parse();
+  ASSERT_TRUE(status.ok()) << status.message();
+  SymbolTable symbol_table(nullptr);
+  const SymbolTableNode& root_symbol(symbol_table.Root());
+
+  const auto build_diagnostics = BuildSymbolTable(src, &symbol_table);
+  EXPECT_EMPTY_STATUSES(build_diagnostics);
```

Comment:
Should we expect a diagnostic on an attempt to re-define `a` in the same scope?

Reply:
We do have such diagnostic (test case covering that is here: https://github.com/google/verible/pull/746/files#diff-45eb889cefc594107d8a95e50aa16ed073b4a7bf296ea244e2c20daa9762fe9cR2558)

I think that you meant an attempt to reference an implictly declared variable.
I thought about that and I think it would be nice to have an extra flag/switch that would report such references but I think that exceeds scope of this PR.